torch/nn/functional.py

```diff
@@ -4761,15 +4761,22 @@ def _in_projection_packed(
     if k is v:
         if q is k:
             # self-attention
-            return linear(q, w, b).chunk(3, dim=-1)
+            proj = linear(q, w, b)
+            # reshape to 3, E and not E, 3 is deliberate for better memory coalescing
+            proj = proj.view(1, *proj.shape[:-1], 3, E).transpose(0, -2).squeeze(-1).contiguous()
```

Comment:
will this make old checkpoints return a different result due to a different interpretation of output channels? or is the interpretation the same?

Reply:
`chunk` was doing the same as (3, E), so checkpoints won't be affected

```
>>> proj = torch.arange(0, 3 * E)
>>> proj
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29])
>>> proj.chunk(3, dim=-1)
(tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]), tensor([10, 11, 12, 13, 14, 15, 16, 17, 18, 19]), tensor([20, 21, 22, 23, 24, 25, 26, 27, 28, 29]))
>>> proj.view(1, *proj.shape[:-1], 3, E).transpose(0, -2).squeeze(-1).contiguous()
tensor([[[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9]],

        [[10, 11, 12, 13, 14, 15, 16, 17, 18, 19]],

        [[20, 21, 22, 23, 24, 25, 26, 27, 28, 29]]])
```

In this case we do it this way because we specifically know we want the chunks to be contiguous. But in most cases it's fine and faster to use `chunk` without calling `contiguous()`. If there were a `contiguous` flag to `torch.chunk` then yes this kind of trick could also be applied there.

Reply:
yeah, maybe `copy=True/False` argument could be introduced for `chunk`/`split`

Reply:
Made the comment more clear.
I think we can leave `chunk()` like it is now, and if the same pattern appears in multiple places we can think about adding a flag later.

Reply:
Also, there might exist an internal method `torch.transpose_copy`, not sure if it's of any use here

Reply:
Do you know where this is implemented? I couldn't find it.

Reply:
Hmm, not sure. Looks like it's some sort of clutch, and not implemented anywhere :(